pkg/kubectl/cmd/login.go

```diff
@@ -0,0 +1,50 @@
+package cmd
+
+import (
+	"io"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+
+	"k8s.io/kubernetes/pkg/client/restclient"
+	"k8s.io/kubernetes/pkg/client/unversioned/clientcmd"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	utilflag "k8s.io/kubernetes/pkg/util/flag"
+)
+
+func NewCmdLogin(f *cmdutil.Factory, pathOptions *clientcmd.PathOptions, out io.Writer) *cobra.Command {
+
+	cmd := &cobra.Command{
+		Use:   "login",
+		Short: "login Log in to the cluster.",
+		Long:  `login obtains the neccesary credentials to make authenticated requests to the cluster. The type of log in depends on the 'auth-provider' type in your configuration, and will likely require some type of user interaction.`,
+
+		Run: func(cmd *cobra.Command, args []string) {
+			err := RunLogin(out, cmd, f, pathOptions, args)
+			cmdutil.CheckErr(err)
+		},
+	}
+
+	return cmd
+}
+
+func RunLogin(out io.Writer, cmd *cobra.Command, f *cmdutil.Factory, pathOptions *clientcmd.PathOptions, args []string) error {
+
+	flags := pflag.NewFlagSet("", pflag.ContinueOnError)
+	flags.SetNormalizeFunc(utilflag.WarnWordSepNormalizeFunc) // Warn for "_" flags
+
+	cmdCliCfg := cmdutil.DefaultClientConfig(flags)
```

Comment:
> @deads2k: Some of what you're describing sounds to me like using kubectl auth login as an alias for kubectl config set users.me.{password|username}. Am I misunderstanding?

I think that would be a good idea.  `kubectl login` would have to be smart enough to handle a partially completed .kube/config file and a completely empty one.  That's part of the reason that `oc login` looks so strange.

Handling all three cases (1. empty config, 2. partial config, 3. re-auth as same or different user) in the single command works very smoothly on the CLI and matches user expectations.

It makes it very easy to start with a .kube/config that a cluster-admin distributed, but also keeps it easy to work from scratch or change your mind.  It's a large part of the reason that we haven't had trouble with difficult to use `kubectl config *` commands.  The primitives are useful, but no user actually uses them outside a script.